### PR TITLE
Support hourly data

### LINKED ISSUE #50: Add support for hourly data
Currently our implementation (encoder_decoder, dataset class and utils) support daily data. To support hourly input data:

- [ ] The encoder_decoder and all model components remain unchanged. The architecture is completely agnostic to whether T=31 or T=744 (24*31). it just sees a temporal dimension to patch. The only model instantiation change is adjusting patch_size (e.g., patch_size=(24, 4, 4) for hourly) and max_days (e.g., max_days=744).
- [ ] The dataset class need changes.
- [ ] The util function `add_month_day_dims` doesnot work for hourly data. We need another function for hourly data.

### PR DESCRIPTION
**Pull request by SarahAlidoost**
closes #50 

This PR:
- adds support for hourly data
- refactor code to improve efficiency 
- adds instructions for dkrz jupyter hub

Issues found:
- dataset should support patching in time to add months to batch dimensions, see #62
- the code should be checked for gpu device and dtype, see #30  ---> draft PR #63

### CODE REVIEW

**Comment by SarahAlidoost** [comment]
@meiertgrootes and @rogerkuou I implemented the support for hourly data and ran the notebook on dkrz jupyter notebook. 

- When using hourly data, the input shape grows quickly e.g. one month: (31×24, 160, 400) = (T, H, W). For longer periods, this becomes too large for memory, even infeasible. Therefore, we cannot train on continuous multi-month sequences. Instead, we should group each month in the batch dimension, e.g. (2, 31×24, 160, 400) = (B, T, H, W) for two months, see issue #62 .
- I refactored parts of the code to reduce CPU data overhead. Training is now faster, but performance is still limited by CPU constraints. We should moving to GPU, but the code must first be debugged for GPU usage, see issue #30 .
-  We could run the "validation" in parallel with the training loop to improve performance, but it is currently not a bottleneck. The inference (forward call) is fast, but the loss.backward in train mode is the challenge. 
- In the example notebook, training was only run for 50 epochs, so the results might not be valid. They should not be used to evaluate the performance of the model in SST prediction.

File: climanet/dataset.py [hunk 1]
```diff
@@ -53,46 +67,55 @@ def __init__(
...
+                daily_da, monthly_da, time_dim=time_dim
+            )
+        else:
+            # Reshape daily → (M, T=31, H, W), monthly → (M, H, W),
+            # and get padded_days_mask → (M, T=31)
+            daily_mt, monthly_m, padded_days_mask, daily_timef = add_month_day_dims(
+                daily_da, monthly_da, time_dim=time_dim
+            )
 
         # Convert to numpy once — all __getitem__ calls use these
```
Comment (meiertgrootes) [inline climanet/dataset.py]: We should probably update tis comment as the conversion format is now a torch tensor rather than numpy

File: climanet/dataset.py [hunk 2]
```diff
@@ -53,46 +67,55 @@ def __init__(
...
-            self.land_mask_np = lm
+            self.land_mask_t = lm
         else:
-            self.land_mask_np = None
+            self.land_mask_t = None
 
         # Precompute the NaN mask before filling NaNs
         # daily_mask: True where NaN (i.e. missing ocean data, not land)
-        self.daily_nan_mask = np.isnan(self.daily_np)  # (M, T=31, H, W)
+        self.daily_nan_mask = torch.isnan(self.daily_t)  # (M, T=31, H, W)
```
Comment (meiertgrootes) [inline climanet/dataset.py]: adding the _t suffix to tensors for clarity as you've done above is a really good idea. Should we adopt that consistently, i.e. also for the daily_nan_mask? Or do you think that is not needed?

File: climanet/predict.py [hunk 3]
```diff
@@ -100,44 +99,46 @@ def predict_monthly_var(
...
             )
 
             # Compute masked loss
             loss = compute_masked_loss(
                 predictions,
-                batch["monthly_patch"].to(device, non_blocking=use_cuda),
-                batch["land_mask_patch"].to(device, non_blocking=use_cuda),
+                batch["monthly_patch"],
+                batch["land_mask_patch"],
             )
```
Comment (meiertgrootes) [inline climanet/predict.py:138]: Using detach and item in this way nicely enables support for cpu and gpu. elegant!

File: climanet/st_encoder_decoder.py [hunk 4]
```diff
@@ -27,16 +24,14 @@ class VideoEncoder(nn.Module):
     https://arxiv.org/abs/2203.12602
     """
```
Comment (meiertgrootes) [inline climanet/st_encoder_decoder.py:26]: You have chosen to get rid of dropout regularization throughout? Could you expand on why in the PR?
This will also propagate to the projection NN used for spatial embeddings. likely not a significant issue.

File: climanet/st_encoder_decoder.py [hunk 5]
```diff
@@ -232,16 +228,13 @@ class TemporalAttentionAggregator(nn.Module):
...
         """Initialize the temporal attention aggregator.
 
         Args:
             embed_dim: Dimension of the embedding. The default is 128.
                 Many vision transformers use embedding dimensions that are multiples
                 of 64 (e.g., 64, 128, 256). This can be tuned.
-            max_days: Maximum length of the temporal dimension to precompute
-            encodings for. Default is 31, which is sufficient for a month of
-            daily data.
             max_months: Maximum number of months (temporal patches) to precompute
```
Comment (meiertgrootes) [inline climanet/st_encoder_decoder.py:238]: We have dropped the day encodings in favor of doy and hod sensitive cyclical encodings. We have kept the explicit month encoding, which will provide varying encodings depending on the length of the sequence in months used and/or will potentially provide the same encodings for different months if a different startig point in the sequnce of months is used.
Should we consider dropping this or/and adding a month-of-year feature to the cyclical encoding module?

File: climanet/st_encoder_decoder.py
Comment (meiertgrootes) [inline climanet/st_encoder_decoder.py:247]: We have dropped the day encodings in favor of doy and hod sensitive cyclical encodings. We have kept the explicit month encoding, which will provide varying encodings depending on the length of the sequence in months used and/or will potentially provide the same encodings for different months if a different startig point in the sequnce of months is used.
Should we consider dropping this or/and adding a month-of-year feature to the cyclical encoding module?

File: climanet/st_encoder_decoder.py [hunk 6]
```diff
@@ -279,10 +272,13 @@ def __init__(self, embed_dim=128, max_days=31, max_months=12, dropout=0.0):
             nn.GELU(),
             nn.Dropout(dropout),
             nn.Linear(4 * embed_dim, embed_dim),
-            nn.Dropout(dropout),
         )
 
-    def forward(self, x, M, T, H, W, time_features, padded_days_mask=None):
+        # Pre-compute and register as buffer — auto-moves with .to(device/dtype)
```
Comment (meiertgrootes) [inline climanet/st_encoder_decoder.py:277]: see comment above

File: climanet/st_encoder_decoder.py [hunk 7]
```diff
@@ -298,41 +294,46 @@ def forward(self, x, M, T, H, W, time_features, padded_days_mask=None):
...
-        temp_emb = temp_emb.expand(-1, H * W, -1, -1, -1)
-        pe_months = self.pos_months(M).to(seq.device).to(seq.dtype)  # (M, C)
+        x = x.reshape(B, Hp, Wp, M, Tp, C)
+        x = x.permute(0, 3, 4, 1, 2, 5)
+        seq = x.reshape(B, M, Tp, HW, C).permute(0, 3, 1, 2, 4)
 
-        seq = seq + temp_emb  # add temporal embeddings
-        seq = seq + pe_months[None, None, :, None, :]  # add month PE
+        temp_emb = self.time_embed(time_features)
+        pe_months = self.pe_months_cache[:M]
```
Comment (meiertgrootes) [inline climanet/st_encoder_decoder.py:304]: see above. In addition, if we move to folding months into the batch dimension, adding a month-of-year to the cyclic time embedding would preserve the information in a simple manner.

File: docs/levante_usage.md [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+sessions.
+
+1. Install uv:
+
+```bash
+wget -qO- https://astral.sh/uv/install.sh | sh
+source $HOME/.local/bin/env
+```
+
+2. Create a new conda environment and install ipykernel and climanet:
```
Comment (meiertgrootes) [inline docs/levante_usage.md]: Technically this isn't a conda environment, but a virtual environment crreated (and managed) by uv.

The reason to be pedantic about this, is that otherwise people might try to use (exisiting) conda installs to modify the environment.

It is possible to use uv with conda, but that requires it's own specific setup (and is not recommended afaik)

**Review by meiertgrootes — APPROVED** [review]
Very nice implementation, including the elegant support for both CPU ad GPU. Some minor comments, but non-blocking for the merge.

In particular, with the envisaged move of months to batch dimension, it would be good to consider dropping the month positional encoding.

Another low priority point (maybe before final relase) is the use of daily_* for the input values and monthly_* for target. the monthly is fine, but the daily can also be hourly. Maybe we can consider renaming somehow. Clearly not crucial

File: climanet/st_encoder_decoder.py (hunk 4 above)
Comment (SarahAlidoost) [reply climanet/st_encoder_decoder.py:26]: Good point. To fix the performance and investigate the bottlenecks, I used the PyTorch profiler with record_function. One thing that stood out was the large number of `aten::dropout and aten::bernoulli` calls, under the class `VideoEncoder`. I originally added `dropout` to help with overfitting. But we are now using `torch.optim.AdamW` and a validation dataset during training. I tested both keeping and removing the dropout in this class. Removing it improved the runtime without affecting the results, so I decided to remove it here. I left the dropout in some other classes because they didn't show up as bottlenecks in the profiler results.

File: climanet/dataset.py (hunk 2 above)
Comment (SarahAlidoost) [reply climanet/dataset.py]: agreed, it is fixed.

File: climanet/dataset.py (hunk 1 above)
Comment (SarahAlidoost) [reply climanet/dataset.py]: fixed.

File: climanet/st_encoder_decoder.py (hunk 5 above)
Comment (SarahAlidoost) [reply climanet/st_encoder_decoder.py:238]: Good point. As you suggested we can implement month-of-year. This issue is related to PR https://github.com/ESMValGroup/ClimaNet/pull/64.

File: climanet/st_encoder_decoder.py (hunk 7 above)
Comment (SarahAlidoost) [reply climanet/st_encoder_decoder.py:304]: correct! it can be fixed in another PR. This is related to #64.

File: docs/levante_usage.md (hunk 8 above)
Comment (SarahAlidoost) [reply docs/levante_usage.md]: correct! it is now fixed.

**Comment by SarahAlidoost** [comment]
> Very nice implementation, including the elegant support for both CPU ad GPU. Some minor comments, but non-blocking for the merge.

@meiertgrootes Hi, thanks for the review and useful comments/suggestions!  I addressed your comments and kept the conversations open. But I will merge this to make the other PRs ready for review too. 

> In particular, with the envisaged move of months to batch dimension, it would be good to consider dropping the month positional encoding.

As you suggested, we can implement this. Related to #64 

> Another low priority point (maybe before final relase) is the use of daily_* for the input values and monthly_* for target. the monthly is fine, but the daily can also be hourly. Maybe we can consider renaming somehow. Clearly not crucial

Right! I fixed a few of them in this PR. But I also made issue #67 to fix the rest.
